### PR TITLE
chore: fix issues raised by clang-tidy (part 2)

### PR DESCRIPTION
**Pull request by xatier**
This PR fixes the following code issues raised by `cert-*,readability-*,`

- simplify nullptr checks
- explicit boolean conversions for length() == 0
- always have parentheses on if/for blocks
- avoid else block after return/break
- use !vector.empty() instead of size() > 0
- use !str.empty() instead of != ""
- const member functions
- fix inconsistent parameter names between header/impl files
- verbose auto for `auto *` and `const auto *` for semantic hints
- mark unused parameters with /*unused*/

- some controversial/noisy readability checks are disabled.

----

@lukhnos, I do have a few remarks/questions:

- there are a few empty `if (false)` blocks in `src/Engine/Mandarin/Mandarin.cpp`, which I wasn't sure the intention, I left them untouched and suppressed the linter.
- There are a few magic numbers in `src/Engine/ParselessPhraseDB.cpp`, can you please share more contexts on these?
- Some "readability rules" are somewhat opinionated, please let me know if you'd like to revert any of them, I can amend the PR.
- all warnings from `clang-tidy` are gone/disabled with this PR, please let me know if you would like to enable `clang-tidy` in CMake by default.

### CODE REVIEW

**Comment by xatier** [comment]
@lukhnos please kindly take a look and answer the above questions. Thanks!

File: src/Engine/Mandarin/Mandarin.cpp
Comment (lukhnos) [inline src/Engine/Mandarin/Mandarin.cpp:80]: Feel free to remove the initial `if (false)`. The code was originally written such that all branches could fit in single lines:
```
if (0) {}
else if (...) { ... }
else if (...) { ... }
```

So that everything could read like a look-up table, and the first `if (0)` was there to "remove the first true branch". But now that we apply clang-format throughout, that is moot.

File: src/Engine/Mandarin/Mandarin.cpp
Comment (lukhnos) [inline src/Engine/Mandarin/Mandarin.cpp:180]: All the `if (false)` occurences can be removed. Same below.

File: src/Engine/ParselessPhraseDB.cpp
Comment (lukhnos) [inline src/Engine/ParselessPhraseDB.cpp:39]: Feel free to remove lines 46-51 (from `uint32_t x = ...` to `assert (x == ...)`, so that you can remove the magic number suppression. These lines have the same effect as lines 41-44 as a way to sanity-check the header, and are actually redundant.

File: src/KeyHandler.cpp
Comment (lukhnos) [inline src/KeyHandler.cpp:51]: Hmm this one is a bit surprising. What if you move the last `return "Standard";` to a final else branch? Would both gcc and clang be happy with it?

**Review by lukhnos — APPROVED** [review]
Thank you so much!

I've answered most of your questions in my review comments. As for:

> Some "readability rules" are somewhat opinionated, please let me know if you'd like to revert any of them, I can amend the PR.

They are fine, and I think having those in place is in general a good thing for code review purposes.

> all warnings from clang-tidy are gone/disabled with this PR, please let me know if you would like to enable clang-tidy in CMake by default.

Just wanted to confirm: it will not attempt to run clang-tidy if it's not available on a local machine, right? If that's the case, I think enabling it by default is a good thing.

**Review by lukhnos — CHANGES_REQUESTED** [review]
Overall LGTM, just that I left a few comments which actually requested some changes. Thanks!

**Comment by xatier** [comment]
Thank you so much! I'll be a bit busy this weekend, will try to make some time to update this PR.

File: src/Engine/Mandarin/Mandarin.cpp
Comment (xatier) [reply src/Engine/Mandarin/Mandarin.cpp]: fixed in https://github.com/openvanilla/fcitx5-mcbopomofo/pull/84/commits/fe162788bb678085edc195ab09e40013eea0f95e

File: src/Engine/Mandarin/Mandarin.cpp
Comment (xatier) [reply src/Engine/Mandarin/Mandarin.cpp]: fixed in https://github.com/openvanilla/fcitx5-mcbopomofo/pull/84/commits/fe162788bb678085edc195ab09e40013eea0f95e

File: src/Engine/ParselessPhraseDB.cpp
Comment (xatier) [reply src/Engine/ParselessPhraseDB.cpp]: fixed in https://github.com/openvanilla/fcitx5-mcbopomofo/pull/84/commits/fe162788bb678085edc195ab09e40013eea0f95e

File: src/KeyHandler.cpp
Comment (xatier) [reply src/KeyHandler.cpp:51]: @lukhnos, I didn't find the rule is applicable here. It would prefer refactoring the following

```cpp
if () {
  return a;
} else if () {
  return b;
} else if () {
  return c;
}
...
```

into

```cpp
if () {
  return a;
}
if () {
  return b;
}
if () {
  return c;
}
...
```

Theoretically, they are identical, but I find this is a bit too lengthy and verbose. We are clearly returning different things w/ given layout.

I believe readability is all about letting the code reader spending minimum mental effort to understand the code intention, which we were clear. My decision is to suppress this rule locally, how do you think?

https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

**Comment by xatier** [comment]
hi @lukhnos, I enabled `clang-tidy` by default and have it skipped if not found. Fixed all the issues mentioned above except the last one, I'd like to see your input.

File: src/KeyHandler.cpp
Comment (lukhnos) [reply src/KeyHandler.cpp:51]: > I believe readability is all about letting the code reader spending minimum mental effort to understand the code intention, which we were clear. My decision is to suppress this rule locally, how do you think?

SGTM. I agree with your assessment about this particular rule.

**Review by lukhnos — APPROVED** [review]
I've tried your branch. Now that clang-tidy is enabled by default, I found it's reporting a few more warnings in Mandarin.cpp, UTF8Helper.cpp, and KeyHandler.cpp. I'm using clang-tidy 11.0.1.

Do you plan to continue working on the warnings, or should we hold off enable-clang-tidy-by-default until all warnings are fixed? I don't want to put the burdens on you—other maintainers can also work on this.

Otherwise this LGTM!

**Comment by lukhnos** [comment]
Edit: since you mentioned earlier that all clang-tidy warnings were fixed with this PR, I wonder if I was using a clang-tidy version that was different from yours, hence the question about whether we should revert to your original setup, which is only to enable clang-tidy if `-DENABLE_CLANG_TIDY=On` is explicitly used when invoking CMake.

**Comment by xatier** [comment]
@lukhnos interesting ... clang 11.1.0 is old [1], I'm on 15.0.7 (archlinux). Arch actually has 16.0.6 released 2 days ago [2], but I haven't updated just yet. Let me revert my change, I would avoid causing confusion to other maintainers with older versions of clang.

```
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 15.0.7
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake
```

https://github.com/llvm/llvm-project/releases/tag/llvmorg-11.1.0
https://archlinux.org/packages/extra-staging/x86_64/clang/

**Comment by xatier** [comment]
As an experiment, I added a CI step to a branch in my forked repo.

clang version installed: https://github.com/xatier/fcitx5-mcbopomofo/actions/runs/5916113158/job/16043026110#step:4:84

```
extra/clang                             15.0.7-9 
```

enable clang-tidy: https://github.com/xatier/fcitx5-mcbopomofo/actions/runs/5916113158/job/16043026110#step:4:312

```
-- Enabled clang-tidy
-- Found clang-tidy: /usr/sbin/clang-tidy
```

As you can see, there's no warning reported by clang-tidy 15.

CI log link: https://github.com/xatier/fcitx5-mcbopomofo/actions/runs/5916113158/job/16043026110

**Comment by lukhnos** [comment]
> Let me revert my change, I would avoid causing confusion to other maintainers with older versions of clang.

I think we can always enable it by default once enough of us are on the lastest toolchain. Thank you so much for your contribution!

**Comment by xatier** [comment]
Sweet, thanks! Sure, we can always enable it by default later.

**Comment by xatier** [comment]
I checked a little deeper with older versions of clang-tidy, it appears that `NOLINTBEGIN` syntax was introduced around clang 14. I'm assuming the warnings you saw with clang-tidy 11 were from a few of the checks that I suppressed locally with `NOLINTBEING`.

I saw those warnings on Debian stable with clang-tidy 13, but not after clang-tidy 14. We should be good to enable this by default after most distros are shipping with clang 14+.

https://reviews.llvm.org/D108560

**Pull request merged**
